### ISSUE TITLE
Closed windows are not removed from the tree on macOS 15 Sequoia

### ISSUE THREAD

**Issue by FelixLisczyk** [nikitabobko/AeroSpace#471]
I'm using Aerospace v0.14.2-Beta on the latest macOS 15 beta (24A5331b). When I close a window, it remains in the tree hierarchy and still takes up part of the screen. This issue does **not** occur on my macOS 14 installation.

# Steps to Reproduce

1. Go to an empty workspace
2. Open Finder
3. Open a second Finder window
4. Close the second Finder window

# Expected Result

The first Finder window should fill the entire display.

# Actual Result

The first Finder window does not expand to fill the entire display.

# Additional Notes

This issue also occurs with other applications.

Here is the output of `aerospace list-windows --workspace visible` after closing the second Finder window:

```
1034 | Finder | 
1028 | Finder | Desktop
```

Console output of `aerospace debug-windows`: [debug-windows.txt](https://github.com/user-attachments/files/16800343/debug-windows.txt)

## Checklist

- [x] I've searched for duplicates. My search queries were (please don't forget to search in closed issues too):
  - macos 15
  - macos sequoia
- [x] I've read the documentation https://nikitabobko.github.io/AeroSpace/guide & https://nikitabobko.github.io/AeroSpace/commands
- [x] I've searched in the documentation. My search queries were:
  - close
  - tree

```
aerospace CLI client version: 0.14.2-Beta 0cb8dbdfc5ee73b8cbc200f175f467ebead55201
AeroSpace.app server version: 0.14.2-Beta 0cb8dbdfc5ee73b8cbc200f175f467ebead55201
```

**Comment by nikitabobko** [comment]
Can't reporduce in `15.1 Beta (24B5035e)`. The problem is somewhere on your side

**Comment by FelixLisczyk** [comment]
@nikitabobko Thank you for testing! I upgraded to macOS 15.1 beta 3 and reinstalled AeroSpace, but the issue continued. I then quit all running apps and background processes, and I found that the [Contexts app](https://contexts.co/) was causing the problem. If anyone else has this issue, quit Contexts and restart AeroSpace to fix it.

**Comment by kennethkn** [comment]
Same issue here. I'm on macOS 15.0 RC (24A335) & AeroSpace v0.14.2-Beta 0cb8dbdfc5ee73b8cbc200f175f467ebead55201. If you repeatedly close and open a finder window, it gets a smaller tile each time, which means windows that are closed aren't getting removed from the tree. I have no other active applications that could cause this. Would appreciate any help to troubleshoot.

**Comment by kennethkn** [comment]
Thanks Felix! I forgot to consider my menu bar apps. After debugging, it was Amazon Q that's causing the issue in my case. But it's worth noting that prior to the OS update, both apps were working in harmony. I noticed that is the case for your app as well. So what could be causing such a conflict? Because not using some app is more of a workaround than a fix. What our apps have in common is that they both install an overlay, which somehow interferes with how Aerospace detects active windows?

**Comment by FelixLisczyk** [comment]
To my knowledge, Aerospace utilizes [Apple's accessibility API](https://developer.apple.com/documentation/applicationservices/axuielement_h) to detect and manage windows. Sequoia has implemented several changes to this API, which has caused problems for window managers like [Amethyst](https://github.com/ianyh/Amethyst/issues/1662). Unfortunately, I am not familiar enough with this API to identify the exact issue. It might be possible to reproduce the problem using a demo app in Xcode by comparing the list of windows reported by the accessibility API before and after closing a window. This could then be reported to the developers of the affected apps. The Contexts app hasn't been updated for several years, so I have now switched to an alternative.

**Comment by FelixLisczyk** [comment]
The yabai window manager experienced the same issue when using Contexts or Amazon Q ([Issue 2324](https://github.com/koekeishiya/yabai/issues/2324#issuecomment-2376590474), [Issue 2431](https://github.com/koekeishiya/yabai/issues/2431#issuecomment-2376560387)). A potential fix has been implemented ([Commit](https://github.com/koekeishiya/yabai/commit/6f9006dd957100ec13096d187a8865e85a164a9b)). Perhaps this workaround could also be applied to AeroSpace, @nikitabobko?

**Comment by haveanicedavid** [comment]
Running into this issue as well - I use both Contexts and Amazon Q. 

Binding a key to the 'close' command (ie `alt-x = 'close'`) works as expected, but `cmd+w` leaves a phantom window:

![CleanShot 2024-10-02 at 13 55 08](https://github.com/user-attachments/assets/31b032dc-5a73-4d70-8f31-25e42f8e1e67)

**Comment by avramukk** [comment]
the same issue for me(
![CleanShot 2024-10-04 at 16 12 08@2x](https://github.com/user-attachments/assets/e240f5ef-5ae3-41d1-920d-fc141724471e)
Beta 15.1

**Comment by nikitabobko** [comment]
> The yabai window manager experienced the same issue when using Contexts or Amazon Q (https://github.com/koekeishiya/yabai/issues/2324#issuecomment-2376590474, https://github.com/koekeishiya/yabai/issues/2431#issuecomment-2376560387). A potential fix has been implemented ([Commit](https://github.com/koekeishiya/yabai/commit/6f9006dd957100ec13096d187a8865e85a164a9b)). Perhaps this workaround could also be applied to AeroSpace, @nikitabobko?

yabai uses private APIs in their commit. I think that it's possible to circumvent this Sequoia bug using public APIs. My solution is described in #445

**Comment by jsonMartin** [comment]
> To my knowledge, Aerospace utilizes [Apple's accessibility API](https://developer.apple.com/documentation/applicationservices/axuielement_h) to detect and manage windows. Sequoia has implemented several changes to this API, which has caused problems for window managers like [Amethyst](https://github.com/ianyh/Amethyst/issues/1662). Unfortunately, I am not familiar enough with this API to identify the exact issue. It might be possible to reproduce the problem using a demo app in Xcode by comparing the list of windows reported by the accessibility API before and after closing a window. This could then be reported to the developers of the affected apps. The Contexts app hasn't been updated for several years, so I have now switched to an alternative.

Hey @FelixLisczyk , I also use currently use Contexts app for MacOS, but it's buggy with AeroSpace. What alternative have you switched to?

**Comment by FelixLisczyk** [comment]
@jsonMartin I'm using AeroSpace and the Raycast window switcher to organize my windows. This issue is currently being worked on in #445, so it may work better with the Contexts app in the future.

**Comment by jsonMartin** [comment]
> @jsonMartin I'm using AeroSpace and the Raycast window switcher to organize my windows. This issue is currently being worked on in #445, so it may work better with the Contexts app in the future.

Thanks! Funny you mention that, I also use Raycast and found the **Switch Windows** command. I migrated over from Contexts, and it works just about as well for application switching. Happy to remove one piece of redundant software from my stack; and it also works much better with AeroSpace.

I combined the above with [AltTab](https://alt-tab-macos.netlify.app/) for thumbnails and am happy with the result!
